### ISSUE TITLE
New search plugin might break in some cases

### ISSUE THREAD

**Issue by squidfunk** [squidfunk/mkdocs-material#3027]
### Contribution guidelines

- [X] I've read the [contribution guidelines](https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md) and wholeheartedly agree

### I've found a bug and checked that ...

- [X] ... the problem doesn't occur with the `mkdocs` or `readthedocs` themes
- [X] ... the problem persists when all overrides are removed, i.e. `custom_dir`, `extra_javascript` and `extra_css`
- [X] ... the documentation does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem

### Description

The new search plugin in Insiders has some trouble with custom HTML and repeating `#` headers. While custom HTML with literal `hx` tags cannot be expected to be indexed as sections (as there's not corresponding anchor link to jump to), it should still build correctly and be linked to the closest parent section.

### Expected behaviour

The search index is not broken, i.e. does not contain duplicate entries.

### Actual behaviour

The search index contains duplicate null-ish entries, which breaks search:

``` json
...
{
  "location": "",
  "text": "",
  "title": ""
},
{
  "location": "",
  "text": "",
  "title": ""
},
{
  "location": "",
  "text": "",
  "title": ""
},
...
```

### Steps to reproduce

Example 1

``` markdown
# A
# B
# C
```

Example 2

``` markdown
# A
<h2>B</h2>
<h3>C</h3>
```

### Package versions

- Python: 3.9
- MkDocs: 1.2.2
- Material: mkdocs-material-7.2.6+insiders-3.0.0

### Configuration

``` yaml
site_name: My Docs
theme:
  name: material
```

### System information

- Operating system: macOS
- Browser: Chrome

**Comment by willingc** [comment]
Yep, ran across this. It worked for the first search and then could not find any matches on subsequent searches.

**Comment by squidfunk** [comment]
@willingc If you have a reproducible example to share, that would be perfect for further testing! I'll try to fix this asap.

**Comment by squidfunk** [comment]
FYI, for the mean-time, if you're running into this problem and want to install the previous version of Insiders until this is resolved, just add the tag to the end of the URL, i.e. `@7.2.6-insiders-2.13.3`:

```
pip install git+https://${GH_TOKEN}@github.com/squidfunk/mkdocs-material-insiders.git@7.2.6-insiders-2.13.3
```

**Comment by lampensau** [comment]
While running through my documentation I've come across another instance of HTML tags breaking the search index. In some places, we used a crude 'hack' (<img width="900") to force Markdown tables to take full width.

A minimal example for the instance described above (everything after the <img> tag isn't correctly included in the search_index.json):
```
# Example

## Search will be working

and `displaying` basic **formatting**

| Color | State | Description |
|:-:|:-|:-|
| Color 1    | Solid | Descriptor 1 |
| Color 2 | Solid | Descriptor 2 <img width="900"> |

## Search is awesome

... but sadly not working anymore.
```

**Comment by squidfunk** [comment]
@lampensau it appears that Python's default HTML parser, which is also used by MkDocs, is not HTML 5 spec-compliant and doesn't see the `img` tag as a self-closing tag, but as an opening tag without a matching closing tag. Since images are filtered form the search index, all remaining contents are nested under `img` and excluded from the index. A quick fix is to add the closing slash: `<img width="900" />`. I'll see how we can switch to a better, HTML5 compliant parser.

**Comment by squidfunk** [comment]
Okay, so I've tested an HTML 5 parser and it looks like this is the way to go, since it produces much more predictable output than Python's default HTML parser which is more or less XML-based. The search plugin should never crash or generate unpredictable output and match exactly what's rendered in the browser, so an HTML 5 parser is definitely the better choice. I'll try to cook up something within the next days.

**Comment by facelessuser** [comment]
Yeah, if you are referring to HTML5lib, it is better but slower. lxml's HTML is probably quicker, but maybe not quite as browser-like with its parsing as HTML5lib. Python's default HTML is adequate...that is all 🙃 .

**Comment by willingc** [comment]
@squidfunk Thanks for the tip. Unfortunately, the doc repo is private. 

What was happening was I could do one successful search after deployment, but every subsequent search responded with Not Found.

**Comment by squidfunk** [comment]
@facelessuser I'm currently testing [`html5_parser`](https://html5-parser.readthedocs.io/en/latest/), it's pretty quick and looks promising.
@willingc I understand. Let's see if the HTML5 parser fixes the issue 😊

**Comment by facelessuser** [comment]
Okay, it looks like it piggybacks off lxml.

**Comment by facelessuser** [comment]
Good find by the way. It seems it can output in various formats, lxml, BeautifulSoup, and standard Python formats.

But, I'm a little wary including it as a dependencies just due to its complicated install requirements: https://html5-parser.readthedocs.io/en/latest/#installation.

**Comment by squidfunk** [comment]
@facelessuser I haven't seen the Windows installation instructions. Meh. Okay, I'll check if we can somehow achieve a lean solution with the current parser that mimics parts of HTML 5 behavior. For example, we don't need to index void elements (`img`, `br`, ...) anyway, so they could just be generally skipped, which solves the problem @lampensau reported.

**Comment by michael-nok** [comment]
I found an additional issue with the new search plugin and page title definitions.  
My file has a YAML meta header that specifies the page title, and the content starts with an H2 "##" header.

The following file causes a build failure:
```
---
title:          Release notes 21.03
---

## Release notes version 21.03
```

Error message:

File "/usr/local/lib/python3.6/site-packages/material/plugins/search/plugin.py", line 79, in create_entry_for_section
"title": "".join(section.title),
TypeError: sequence item 0: expected str instance, float found

When I change the ## header to #, the build works fine.

**Comment by lampensau** [comment]
Another example for breaking search: Having a link in a Markdown heading will result in the link text missing from the title.

```
# Example

## Search will be working

and `displaying` basic **formatting**.

## Search is [broken](#)

... but will "work" if the title is not completely empty.
```

search_index.json:
```
{"location":"#search-is-broken","text":"<p>... but sadly not working anymore.</p>","title":"Search is "}]}
```

**Comment by squidfunk** [comment]
I fixed all reported cases in the latest 4 commits on Insiders:

- [`ada5441`](https://github.com/squidfunk/mkdocs-material-insiders/commit/ada54410821e194c6faf19f3167cf40167c75dc9) Fixed search plugin breaking on void elements without slashes
- [`1ecf672`](https://github.com/squidfunk/mkdocs-material-insiders/commit/1ecf672076891fadfefc9e6c14a0d941ed96c74a) Fixed search plugin filtering link contents from headlines
- [`af6da3c`](https://github.com/squidfunk/mkdocs-material-insiders/commit/af6da3cb7a41c2e5097118904817e487af7915c3) Fixed search plugin ignoring subsequent `h1` headlines
- [`83050f0`](https://github.com/squidfunk/mkdocs-material-insiders/commit/83050f014a86136ce516100ea58c635ac98527bc) Fixed search plugin not inserting missing `h1` headline

I've decided to not switch to the HTML 5 parser for now, but stick to the current implementation for reasons of performance and ease of installation as pointed out by @facelessuser. Following is a list of examples that are (IMO) now indexed correctly:

### Void elements without slashes

As noted before, this was related to the HTML parser reading this as an opening tag, not a void element. This is now handled correctly by the parser, which will now ignore void elements.

Input:

```` markdown
# Void elements

void elements <img width="900"> should not ...

## 2nd headline

... interrupt the flow.
````

Output:

``` json
{
  "location": "",
  "text": "<p>void elements  should not ...</p>",
  "title": "Void elements"
},
{
  "location": "#2nd-headline",
  "text": "<p>... interrupt the flow.</p>",
  "title": "2nd headline"
}
```

### Link contents in headlines

The new search plugin has a slightly different behavior when it comes to headlines. The old implementation will just take the title from the table of contents extension and set that as the section title, which means that the title cannot contain markup like inline code. The new implementation will take the title exactly as defined from the Markdown, to allow for inline code formatting. However, the problem this exposed was that when permalinks were activated, the permalink content was now part of the title, e.g. `Headline ¶`. Thus, I've decided to just filter the contents of `a` elements when within an `h1-6` element, which was of course, too short-sighted.

This is now handled more granularly, i.e. only permalinks are filtered.

Input:

```` markdown
# Links

## Search is [broken](#)

... but will "work" if the title is not completely empty.

## [Another link](#)

... with content
````

Output:

``` json
{
  "location": "",
  "text": "",
  "title": "Links"
},
{
  "location": "#search-is-broken",
  "text": "<p>... but will \"work\" if the title is not completely empty.</p>",
  "title": "Search is broken"
},
{
  "location": "#another-link",
  "text": "<p>... with content</p>",
  "title": "Another link"
}
```

### Main headline handling (bugs with `h1` headlines)

Subsequent `h1` headlines are now handled correctly, as are missing `h1` headlines, for which a new section will be automatically generated. This is necessary to adhere to the protocol of the frontend search index, which expects the first entry for a page to be the main entry, i.e. page headline. This is basically the same, as when there's a preface before the first `h1`, which will also result in an automatically generated section.

Input:

```` markdown
---
title: My page title
---

Preface

# A

Foo

# B

Bar

# C

Baz
````

Output:

``` json
{
  "location": "",
  "text": "<p>Preface</p>",
  "title": "My page title"
},
{
  "location": "#a",
  "text": "<p>Foo</p>",
  "title": "A"
},
{
  "location": "#b",
  "text": "<p>Bar</p>",
  "title": "B"
},
{
  "location": "#c",
  "text": "<p>Baz</p>",
  "title": "C"
}
```

I'd consider this issue resolved, and will issue a bugfix release in a few hours (after doing some further cleanup). If any new problems arise, please open new issues with reproducible cases. Those cases helped a lot in finding the relevant parts of the code and fixing them quickly. Many thanks to @lampensau and @michael-nok!

**Comment by squidfunk** [comment]
... what's also cool is that now, literal headlines can also induce new sections if they define an `id` attribute. This can be mixed with regular Markdown as desired, which allows for __better integration with auto-generated markup__.

Input:

```` markdown
# Headline

<h2 id="my-custom-id">Subheadline 1</h2>

Some content

<p>More content</p>

## Subheadline 2

And even more
````

Output:

``` json
{
  "location": "",
  "text": "",
  "title": "Headline"
},
{
  "location": "#my-custom-id",
  "text": "<p>Some content</p> <p>More content</p>",
  "title": "Subheadline 1"
},
{
  "location": "#subheadline-2",
  "text": "<p>And even more</p>",
  "title": "Subheadline 2"
}
```

**Comment by squidfunk** [comment]
Released as part of 7.2.7+insiders-3.0.1.
